modio.BufferedWriter: Fix the case where write() fails.

jepler opened the pull request:
### Summary

Previously, there was no test coverage of the "write failed" path. In fact, the assertion would fire instead of gracefully raising a Python exception.

Slightly re-organize the code to place the assertion later. Add a test case, and update the expected output.

### Testing

I ran the unix coverage tests locally.

### Trade-offs and Alternatives

I considered writing `assert(err || (out_sz == self->len));` but thought this was preferable.

jepler committed:
modio.BufferedWriter: Fix the case where write() fails.

Previously, there was no test coverage of the "write failed"
path. In fact, the assertion would fire instead of gracefully
raising a Python exception.

Slightly re-organize the code to place the assertion later.
Add a test case, and update the expected output.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented on py/modio.c:
```diff
@@ -169,12 +169,13 @@ static mp_obj_t bufwriter_flush(mp_obj_t self_in) {
         if (err != 0) {
             mp_raise_OSError(err);
+        } else {
+            // TODO: try to recover from a case of non-blocking stream, e.g. move
```
I think the comment is referring to the case where e.g., 8 bytes were requested to write, but only 6 bytes were written. This is the case that hits the assertion error, and where moving the "remaining" bytes to the head of the buffer would be a possible way to recover. If that understanding is right, then I think the comment's in the right place.

 * err == 0, out_sz == self->len: data was written & flushed
 * err != 0: data is preserved, OSError is raised (what this PR implements)
 * err == 0, out_sz < self->len: data is partially written. assertion will fire. If assertions disabled, some data is lost.

jepler commented on tests/basics/io_buffered_writer.py:
```diff
@@ -28,3 +28,19 @@
+    def write(self, buf):
+        return None
+
+    def writable(self):
```
you're right, will change. I had toyed with *NOT* printing the value on line 17 and getting rid of the exp file; this is a remnant of that, I guess. Will remove.

dpgeorge commented on tests/basics/io_buffered_writer.py:
```diff
@@ -28,3 +28,16 @@
+    buf.flush()
+    print("flushed")
+except OSError:
+    print("OSError")
```
I think this needs to test a little more, to ensure we have the logic correct for the case of a failed flush.

In particular we want to test that `self->len = 0` should not be executed if there's an error.  Because if we turn the assertion off then the change here changes semantics.

I suggest to loop over this try-except 4 times:
- the first two should fail with errors and print OSError
- the third should succeed and flush all 6 bytes (and `MyIO.write` should print those bytes out)
- the fourth should succeed and flush nothing, and not even call `MyIO.write`

To do this you can add some simple state to `MyIO` so it fails the first two calls to `write`, then succeeds the remaining calls.

jepler committed:
io_buffered_writer.py: Improve failed flush test.

Try flushing 4 times.

The first two times, the underlying IO object fails.
The third time, the data is actually flushed.
The last time, write is not called as there was no data left.

Signed-off-by: Jeff Epler <jepler@gmail.com>

dpgeorge reviewed (approved):
Yes, that's exactly what I was thinking.  Looking good now!

dpgeorge commented:
Squashed and merged in b6b7d64bd9257cc1c138d37033108dcff5ceb89e
